### PR TITLE
feature add Backup Gateway for new design API

### PR DESCRIPTION
**Pull request by kpango**
### Description:
Implement backup gateway for new design API
this PR is sub PR for #742

### Environment:

- Go Version: 1.15.2
- Docker Version: 19.03.8
- Kubernetes Version: 1.18.2
- NGT Version: 1.12.1

### Types of changes:

- [ ] Bug fix [type/bug]
- [ ] New feature [type/feature]
- [ ] Add tests [type/test]
- [ ] Security related changes [type/security]
- [ ] Add documents [type/documentation]
- [ ] Refactoring [type/refactoring]
- [ ] Update dependencies [type/dependency]
- [ ] Update benchmarks and performances [type/bench]
- [ ] Update CI [type/ci]

### Changes to Core Features:

- [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your core changes, as applicable?
- [ ] Have you successfully ran tests with your changes locally?

### Checklist:

- [ ] I have read the [CONTRIBUTING](https://github.com/vdaas/vald/blob/master/CONTRIBUTING.md) document.
- [ ] I have checked open [Pull Requests](https://github.com/vdaas/vald/pulls) for the similar feature or fixes?
- [ ] I have added tests and benchmarks to cover my changes.
- [ ] I have ensured all new and existing tests passed.
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: pkg/gateway/backup/usecase/vald.go [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+			case <-ctx.Done():
+				return ctx.Err()
+			case err = <-oech:
+			case err = <-bech:
+			case err = <-sech:
+			}
+			if err != nil {
+				select {
+				case <-ctx.Done():
+					return ctx.Err()
```
Comment (hlts2) [inline pkg/gateway/backup/usecase/vald.go:189]: do you need to returns the wrapped error or to output error log?

File: pkg/gateway/backup/usecase/vald.go [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+	ech := make(chan error, 6)
+	var bech, sech, oech <-chan error
+	var err error
+	if r.observability != nil {
+		oech = r.observability.Start(ctx)
+	}
+	if r.backup != nil {
+		bech, err = r.backup.Start(ctx)
+		if err != nil {
+			close(ech)
```
Comment (hlts2) [inline pkg/gateway/backup/usecase/vald.go:171]: do you need to close oech?

File: pkg/gateway/backup/handler/grpc/handler.go [hunk 3]
```diff
@@ -0,0 +1,705 @@
...
+
+func (s *server) Search(ctx context.Context, req *payload.Search_Request) (res *payload.Search_Response, err error) {
+	ctx, span := trace.StartSpan(ctx, apiName+".Search")
+	defer func() {
+		if span != nil {
+			span.End()
+		}
+	}()
+	if len(req.Vector) < 2 {
+		return nil, errors.ErrInvalidDimensionSize(len(req.Vector), 0)
```
Comment (rinx) [inline pkg/gateway/backup/handler/grpc/handler.go:84]: is it better to write like this?
```suggestion
	if len(req.GetVector()) < 2 {
		return nil, errors.ErrInvalidDimensionSize(len(req.GetVector()), 0)
```

File: pkg/gateway/backup/handler/grpc/option.go [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+func WithBackup(b service.Backup) Option {
+	return func(s *server) {
+		if b != nil {
+			s.backup = b
+		}
+	}
+}
+
+func WithClient(g vald.Client) Option {
```
Comment (rinx) [inline pkg/gateway/backup/handler/grpc/option.go:43]: same as the other PR, it is better to name it as WithValdClient.

File: pkg/gateway/backup/router/router.go [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+				},
+				"/search/multi",
+				h.MultiSearch,
+			},
+			{
+				"Multi Search By ID",
+				[]string{
+					http.MethodGet,
+				},
+				"/search//multi/{id}",
```
Comment (rinx) [inline pkg/gateway/backup/router/router.go:82]:
```suggestion
				"/search/multi/{id}",
```

File: pkg/gateway/backup/service/backup.go [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+	}
+
+	return b, nil
+}
+
+func (b *backup) Start(ctx context.Context) (<-chan error, error) {
+	return b.client.StartConnectionMonitor(ctx)
+}
+
+func (b *backup) GetObject(ctx context.Context, uuid string) (vec *payload.Backup_MetaVector, err error) {
```
Comment (rinx) [inline pkg/gateway/backup/service/backup.go:59]: Could you please add trace spans to GetObject, GetLocation, etc...

File: .github/workflows/dockers-gateway-backup-image.yml [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+      with:
+        registry: ghcr.io
+        username: ${{ secrets.PACKAGE_USER }}
+        password: ${{ secrets.PACKAGE_TOKEN }}
+    - name: Image name
+      id: image_name
+      run: |
+        image_name=`make docker/name/gateway-backup`
+        base_platforms=`make docker/platforms`
+        echo "::set-env name=IMAGE_NAME::${image_name}"
```
Comment (rinx) [inline .github/workflows/dockers-gateway-backup-image.yml:85]:
```suggestion
        echo "IMAGE_NAME=${image_name}" >> $GITHUB_ENV
```

File: .github/workflows/dockers-gateway-backup-image.yml [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+        echo "::set-output name=IMAGE_NAME::${image_name}"
+        echo "::set-output name=BASE_PLATFORMS::${base_platforms}"
+    - name: Determine tag name (master)
+      if: github.ref == 'refs/heads/master'
+      run: |
+        echo "::set-env name=PRIMARY_TAG::nightly"
+        echo "::set-env name=PLATFORMS::${PLATFORMS}"
```
Comment (rinx) [inline .github/workflows/dockers-gateway-backup-image.yml:92]:
```suggestion
        echo "PRIMARY_TAG=nightly" >> $GITHUB_ENV
        echo "PLATFORMS=${PLATFORMS}" >> $GITHUB_ENV
```

File: .github/workflows/dockers-gateway-backup-image.yml [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+      env:
+        PLATFORMS: linux/amd64,linux/arm64
+    - name: Determine tag name (pull request)
+      if: github.event_name == 'pull_request'
+      run: |
+        pr_num=`cat $GITHUB_EVENT_PATH | jq -r ".number"`
+        echo "PR-${pr_num}" > versions/VALD_VERSION
+        echo "::set-env name=PRIMARY_TAG::pr-${pr_num}"
+        echo "::set-env name=PLATFORMS::${PLATFORMS}"
```
Comment (rinx) [inline .github/workflows/dockers-gateway-backup-image.yml:101]:
```suggestion
        echo "PRIMARY_TAG=pr-${pr_num}" >> $GITHUB_ENV
        echo "PLATFORMS=${PLATFORMS}" >> $GITHUB_ENV
```

File: .github/workflows/dockers-gateway-backup-image.yml [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+      env:
+        PLATFORMS: linux/amd64
+    - name: Determine tag name (tags)
+      if: startsWith( github.ref, 'refs/tags/')
+      id: determine_tag
+      run: |
+        tag_name=`echo $GITHUB_REF | sed -e 's:^refs/tags/::'`
+        echo "::set-output name=TAG_NAME::${tag_name}"
+        echo "::set-env name=PRIMARY_TAG::${tag_name}"
+        echo "::set-env name=PLATFORMS::${PLATFORMS}"
```
Comment (rinx) [inline .github/workflows/dockers-gateway-backup-image.yml:111]:
```suggestion
        echo "PRIMARY_TAG=${tag_name}" >> $GITHUB_ENV
        echo "PLATFORMS=${PLATFORMS}" >> $GITHUB_ENV
```

**Review by rinx — COMMENTED** [review]
These changes are related to #752

File: pkg/gateway/backup/usecase/vald.go (hunk 1 above)
Comment (kpango) [reply pkg/gateway/backup/usecase/vald.go:189]: I don't think so

**Comment by kpango** [comment]
@rinx @hlts2 resolved.

**Comment by rinx** [comment]
maybe it's okay.

**Review by hlts2 — APPROVED** [review]
LGTM

**Pull request merged**